### PR TITLE
Error for negative start index in string section_issues#4871

### PR DESCRIPTION
**Pull request by AnshMehta1**
This pull request aims to fix issue https://github.com/lfortran/lfortran/issues/4871. Basically , a compile-time check is done for negative start index in case of character array splicing or substring section and correspondingly , a semantic error is thrown for negative start index. Changes in codebase are made accordingly to check for character array and then for start index.

### CODE REVIEW

**Comment by Pranavchiku** [comment]
Awesome, you figured it out!

**Comment by AnshMehta1** [comment]
No , actually I ignored the condition where start index is 0. Actually, I am not sure how to deal that.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -6912,6 +6912,19 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
+            
+                if (x.m_args != nullptr && x.m_args[0].m_start != nullptr) {
+
+                    if((x.m_args[0].m_start->type) == AST::exprType::UnaryOp){
+                        throw SemanticError("The first index in string section is negative", x.base.base.loc);
+                    }
+                }
+            }
+        }
+
```
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_common_visitor.h:6927]: Maybe you can follow the diff and try to understand if it makes sense. I am not sure if the diff is a complete fix but will give you directions:

```diff
diff --git a/src/lfortran/semantics/ast_common_visitor.h b/src/lfortran/semantics/ast_common_visitor.h
index a76b5cb79..d4b109f38 100644
--- a/src/lfortran/semantics/ast_common_visitor.h
+++ b/src/lfortran/semantics/ast_common_visitor.h
@@ -4028,6 +4028,7 @@ public:
                 this->visit_expr(*(m_subargs[0].m_step));
                 step = ASRUtils::EXPR(tmp);
                 step = CastingUtil::perform_casting(step, int_type, al, loc);
+                std::cout<<"here 2"<<std::endl;
                 return ASR::make_StringSection_t(al, loc, array_item, l,
                         r, ASRUtils::EXPR(tmp), char_type, arr_ref_val);
             } else {
@@ -4061,6 +4062,18 @@ public:
                                             ASRUtils::expr_value(args[0].m_left))->m_n - offset;
                             a_value = ASRUtils::EXPR((ASR::make_IntegerConstant_t(al, loc,
                                                     a, int_type)));
+                        } else if ( ASR::is_a<ASR::IntegerUnaryMinus_t>(*args[0].m_left) ) {
+                            std::cout<<"IUM"<<std::endl;
+                            ASR::expr_t* value = ASRUtils::expr_value(*args[0].m_left);
+                            if ( value ) {
+                                int64_t a = -ASR::down_cast<ASR::IntegerConstant_t>(value)->m_n - offset;
+                                if ( a < 0 ) {
+                                    throw SemanticError("Negative index in string slice", loc);
+                                } else {
+                                    a_value = ASRUtils::EXPR((ASR::make_IntegerConstant_t(al, loc,
+                                                    a, int_type)));
+                                }
+                            }
                         }
                         ASR::expr_t* casted_left = CastingUtil::perform_casting(args[0].m_left, int_type, al, loc);
                         l = b.Sub(casted_left, const_1, a_value);
@@ -4085,7 +4098,7 @@ public:
                         char_type = ASRUtils::TYPE(ASR::make_Character_t(al, loc,
                                         1, a_len, a_len_expr, ASR::string_physical_typeType::PointerString));
                     }
-
+                    std::cout<<"here 3"<<std::endl;
                     return ASR::make_StringSection_t(al, loc, v_Var, l,
                             r, casted_step, char_type, arr_ref_val);
                 }
@@ -6919,7 +6932,7 @@ public:
                 if (x.m_args != nullptr && x.m_args[0].m_start != nullptr) {
 
                     if((x.m_args[0].m_start->type) == AST::exprType::UnaryOp){
-                        throw SemanticError("The first index in string section is negative", x.base.base.loc);
+                        // throw SemanticError("The first index in string section is negative", x.base.base.loc);
                     }
                 }
             }
@@ -7190,6 +7203,7 @@ public:
             switch (f2->type) {
             case(ASR::symbolType::Variable): {
                 // TODO: Make create_StringRef for character (non-array) variables.
+                std::cout<<"here 1"<<std::endl;
                 tmp = create_ArrayRef(x.base.base.loc, x.m_args, x.n_args,
                                       x.m_subargs, x.n_subargs, v_expr, v, f2);
                 break;
```

**Comment by AnshMehta1** [comment]
Ok , I will look into it and try to fix it.

**Comment by Pranavchiku** [comment]
```fortran
pure recursive subroutine introsort( array, depth_limit )
! It devolves to `insertionsort` if the remaining number of elements
! is fewer than or equal to `INSERT_SIZE`, `heapsort` if the completion
! of the `quicksort` is too slow as estimated from `DEPTH_LIMIT`,
! otherwise sorting is done by a `quicksort`.
            character(len=*), intent(inout) :: array(0:)
            integer(int32), intent(in)     :: depth_limit

            integer(int_size), parameter :: insert_size = 16_int_size
            integer(int_size)            :: index

            integer(int_size) :: array_len

            array_len = size(array, kind=int_size)

            if ( array_len <= insert_size ) then
                ! May be best at the end of SORT processing the whole array
                ! See Musser, D.R., “Introspective Sorting and Selection
                ! Algorithms,” Software—Practice and Experience, Vol. 27(8),
                ! 983–993 (August 1997).

                call insertion_sort( array )
            else if ( depth_limit == 0 ) then
                call heap_sort( array )
            else
                call partition( array, index )
                call introsort( array(0:index-1), depth_limit-1 )
                call introsort( array(index+1:array_len), depth_limit-1 )
            end if

        end subroutine introsort
```

**Comment by certik** [comment]
Closing in favor of https://github.com/lfortran/lfortran/pull/4976.

**Pull request closed without merging**